File: 39.combination-sum.java
// @leet start
class Pair {
    public int sum;
    public List<Integer> arr;
    Pair(int start) {
        this.sum = start;
        this.arr = new ArrayList<>();
        this.arr.add(start);
    }

    Pair(Pair p) {
        this.sum = p.sum;
        this.arr = new ArrayList<>(p.arr);
    }
}


class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();

        for (int i = 0; i < candidates.length; i++) {
            for (int j = i; j < candidates.length; j++) {
                List<Pair> vals = new ArrayList<>();
                vals.add(new Pair(candidates[i]));
                result.addAll(this.combinationSumHelper(candidates, j, target, vals));
            }
        }

        return result;
    }

    private List<List<Integer>> combinationSumHelper(int[] candidates, int idx, int target, List<Pair> vals) {
        List<Pair> local = new ArrayList<>();
        List<List<Integer>> localResult = new ArrayList<>();
        int val = candidates[idx];

        for (Pair p : vals) {
            if (p.sum + val == target) {
                p.arr.add(val);
                localResult.add(p.arr);
                continue;
            }
            if (p.sum + val > target) {
                continue;
            }
            Pair temp = new Pair(p);
            p.sum += val;
            p.arr.add(val);
            local.add(temp);
        }

        for (int i = idx; i < candidates.length; i++) {
            localResult.addAll(this.combinationSumHelper(candidates, i, target, local));
        }

        return localResult;
    }
}
// @leet end

File: 39.combination-sum.py
# @leet start
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []

        def dfs(i, curr, total):
            if total == target:
                result.append(curr.copy())
                return
            elif i >= len(candidates):
                return
            elif total > target:
                return

            curr.append(candidates[i])
            dfs(i, curr, total + candidates[i])
            curr.pop()
            dfs(i + 1, curr, total)

        dfs(0, [], 0)
            
        return result
# @leet end
